Joint trajectory controller

adolfo-rt opened the pull request:
This is a fully functional and tested implementation of a Joint Trajectory Controller.

**Note:** I developed this controller in my Fuerte backport of ros_controllers, and tried to make this PR Hydro-friendly, but I've not tested it in a catkin system. I suggest that this PR should be merged when someone using Groovy or Hydro is willing to iron out any pending wrinkles resulting from the Hydro port. It should suffice to run the test suite to ensure that everything is working as expected.
### ROS API
- Exposes two ROS APIs for sending commands `FollowJointTrajectory` action and `trajectory_msgs::JointTrajectory` topic.
- Current controller state is available in a published `control_msgs::JointTrajectoryControllerState` topic.
- Controller state at any future time can be queried through a service of type `control_msgs/JointTrajectoryControllerState`.
### Segment type
- Controller is templated on the segment type, so any segment implementation conforming to the expected interface can be used.
- Controllers for new segment types are a template specialization away.
- Multi-dof quintic spline segment implementation provided by default.
### Hardware interface type
- Controller is templated on the hardware interface type.
- Any hardware interface can be supported as long as a template specialization for the `HardwareInterfaceAdapter` exists.
- Controller currently can work with position and effort (untested) hardware interfaces.
### Other
- Realtime-safe.
- Proper handling of wrapping (continuous) joints.
- Discontinuous system clock changes do not cause discontinuities in the execution of already queued trajectory segments.
- Complete test suite and developer documentation.
### TODO
- User-oriented documentation.

Commit:
Layout skeleton if JointTrajectoryController.

Commit:
Add basic trajectory interface and utilities.

- Implementation of single-dof quintic spline segment.

- Multi-dof segment templated on the single-dof segment type.

- General(ish) representation of trajectories, templated on segment type.

Commit:
Initialize segments with start and end times.

Before the segment duration was provided and zero start time was assumed.
Now initializing and sampling segments have a more consistent API:
- You initialize a segment from start and end time-state pairs.
- You sample a segment with a time[in] and a state[out].

Commit:
Add doc configuration file.

Commit:
Add unit tests for segment sampling classes.

Commit:
First trajectory interface functions with tests.

- Find the segment associated to a specified time instant.
- Sample a trajectory at a specified time instant.

Commit:
Doc fixes.

Commit:
Remove dead code.

Commit:
Test trajectories with overlapping segments.

Excercise the case when a segment's end_time is greater than the next segment's
start_time.

Commit:
Trivial doc fix.

Commit:
Harmonize MultiDofSegment constructor signature.

Segment implementations should have a similar way to be constructed, namely
by specifying four parameters:

  start_time, start_state, end_time, end_state

The types of these parameters depends on the segment type, but the structure
is preserved.

Commit:
Add a non-iterator based findSegment(...) overload.

Used for convenience purposes.

Commit:
Add utilities for reading trajectories from ROS msgs.

Commit:
Move trajectory_interface to a separate directory.

General-purpose trajectory_interface functionality should live in a separate
place (ie. ROS package) than the joint_trajectory_controller. This is a first
step towards this separation.

Commit:
Restructure error reporting log statements.

Commit:
Scope init method as protected.

Commit:
Add new trajectory segment type.

- Multi-dof quintic spline segment that can be constructed from ROS message
  data structures.

Commit:
Implement trajectory init from ROS message data.

Update unit tests accordingly.

Commit:
Documentation and log message fixes.

Commit:
findSegment overload returning non-const iterator.

Commit:
Better debug log statements.

Commit:
Trivial doc fix.

Commit:
Add missing bits to conform with Segment API.

Commit:
Draft implementation of splicer, controller update.

- Splicer code might be factored out soon, this is just a test.

Commit:
Implement init() method, except ROS API setup.

Commit:
Add tests for non-ros segment constructor.

Commit:
Add joint reordering on ROS message traj init.

- When creating joint trajectories from ROS messages, it is now (optionally)
  possible to provide a vector of expected joint names, so expected joint count and
  names can be checked. Also, resulting trajectories are ordered as the expected
  joints vector, and not like the ROS message.

- Update unit tests.

Commit:
Add JointTrajectory segment construction options.

- New optional parameter: Permutation vector, useful when joints in ROS message
  are not ordered as the controller expects them.

- New optional parameter: Position offset, useful for handling joints that wrap
  around (ie. continuous), to compensate for multi-turn offsets.

- Propagate changes to user classes and tests. Position offsets are not fully
  integrated yet.

Commit:
Trivial log statement fixes.

Commit:
Reimplement permutation() leveraging STL more.

Commit:
Add size() method.

Commit:
Refactor initJointTrajectory()

- Rename from init().
- Move to a separate header, combine with code in controller.
- Complete wrapping joint support, unit test it.
- The unit test of initJointTrajectory() is commented-out. Needs to be updated.

Commit:
Complete feature set of initJointTrajectory()

- Implement support for wrapping joints.
- Some function parameters are now optional.
- Comprehensive debug log statements.
- Update unit test suite.

Commit:
change initJointTrajectory() signature.

- Instead of having 5 parameters: 2 madatory and 3 optional, now there are
  3 parameters: the same 2 mandatory, and an Options class with optional
  data. Specifying the options is optional.

- Update unit tests.

Commit:
Use uniform naming for wraparound variables.

Commit:
Unit test case where wraparound spec is ignored.

Commit:
Add documentation.

Commit:
Use resize and [] instead of push_back on init.

Commit:
Topic interface support, partial action support.

- Listening to command topics implemented.

- Listening to action goals is implemented, but goal completion (success or
  aborted) is not implemented yet. Requires joint tolerances to be in place.

- Add missing bits to make the plugin load.

Commit:
QuinticSplineSegment is now inherently multi-dof.

- MultiDofSegment wrapper is no longer needed.

- Position, velocity, acceleration aata is now stored in a SoA structure, which is
  easier to manipulate.

Commit:
Complete action interface support.

- Goal completion is checked.

- Tolerance checking implemented, needs cleanup, doc and testing though.

- Reading tolerance data from goal handle and ROS param server is still TODO.

Commit:
Fix current action goal resetting.

Commit:
Move goal handle status setting to controller.

- Basic state constraint validation remains in segment header file.

- Goal handle status management now lives in controller.

- Document constraint validation classes/functions.

Commit:
Better printing of time values.

Commit:
Move controller-specific code to its namespace.

...and out of the trajectory_interface namespace.

Commit:
Trivial doc addition.

Commit:
Trivial doc fix.

Commit:
Rename file for more consistent naming.

Commit:
Deprecate multi_dof_segment.

- Its functionality has been introduced in the QuinticSplineSegment rework that
  made it inherently multi-dof.

Commit:
Add goal handle value checking to init tests.

Commit:
Rename constraints -> tolerances in code.

Commit:
Unit-test trajectory state tolerance checking.

Commit:
Expose Scalar type as segment typedef.

Commit:
Scalar type fully templated accross the board.

To make this happen, and good for other reasons as well: separate State in
QuinticSplineSegment to a separate file.

Commit:
Add missing \endcode in doc.

Commit:
init method is now part of the Segment public API.

Commit:
Make starting() method realtime-safe.

Commit:
Drop usage of system clock in controller loop.

Internal trajectory representation is no longer parameterized on the system
clock, but on a monotonically increasing variable representing controller uptime
(the base is irrelevant, the important thing is that it's monotonically
increasing).

This solves the problem that if the system time changes during controller
operation, the result is a potentially dangerous discontinuity in the joint
commands. It's true that one should not perform abrupt time changes during
operation, and only do very small corrections (eg. NTP slew), but it stands
to reason (and to the principle of least surprise) that already queued
commands should not be affected by an external influence like a time change.

When a new trajectory command arrives, it will be immediately transformed to the
monotonic reference. In this way, system time changes affect only new commands
arriving _after_ the change, and not already queued ones.

Commit:
Protect shared time data behind a realtime buffer.

Note: We're using the realtime_tools::RealtimeBuffer the wrong way around,
ie. we're writing from a realtime thread, and reading from a non-realtime
thread. The ideal solution would be to use a lock-free data structure.

Commit:
Unit test trajectory init with different time base.

- Test initialization from ROS message when current trajectory and message
  are represented in different time bases.

- Fix bug in implementation exposed by the test :)

Commit:
Implement monitoring ROS API.

- Add the "query_state" ROS service.

- Add the "state" topic, whose publish rate is configurable.

- Add checks that either error-out or do a no-op when the ROS API is excercised
  in a stopped state. This is important because the ROS API is available when
  the controller is initialized, but not yet running.

- Label realtime and non-realtime methods on public class API.

Commit:
Move tolerances to separate header.

- Fetch default tolerances from ros param server in controller.

- Document tolerance-realted functions.

Commit:
Unit test tolerances-related functionality.

Commit:
Templatize controller on HW iface & segment types.

- JointTrajectoryController is now a template class that depends on the
  Hardware interface type, as well as on the trajectory segment type. This
  allows to reuse the code across multiple combinations of these two parameters.
  We currently offer plugins for quintic spline segments and position or
  effort hardware interfaces.

- Create a HardwareInterface Adapter class, that converts desired+error states
  from the trajectory sampler (pos, vel, acc) to commands of a given hardware
  interface. Currently implemented adapters for position and effort interfaces.

Commit:
Better logging messages

- Use NAMED log statements, to differentiate individual controllers.

- More informative info message on controller init: No. of joints, hw interface
  and segment type are reported.

Commit:
Add return value to updateTrajectoryCommand method

- Leverage this to discard invalid trajectory messages (eg. too old) and not
  preempt active goals when such invalid messages arrive.

Commit:
HardwareInterfaceAdapter now also sets commands.

Commit:
Trivial cosmetic fixes.

Commit:
Initial controller rostest.

- Currently only exercises topic interface.

Commit:
Add minimal controller doc.

Commit:
First version of catkin-spcific scripts. Untested.

Commit:
Log exception string when traj update fails.

Commit:
Handle singularity at pi when computing wraparound.

- angles::shortest_angular_distance() has a singularity at pi, that can yield
  unexpected results when computing wraparound values. This is now taken into
  account.

- Update unit tests.

Commit:
Remove unused variable.

Commit:
Extend controller rostest.

- Pretty much all of the controller code is exercised, except for tolerance
  checking.

Commit:
Fix bug where goal handle was not being reset.

- When an action goal failed due to path constraint violations, the currently
  active goal was not being reset.

Commit:
Trivial log message fix.

Commit:
Restore use of INVALID_JOINTS error code.

- When rejecting a goal for invalid joints reasons, use INVALID_JOINT, and not
  INVALID_GOAL.

Commit:
Allow to tune robot hardware simulator fidelity.

- Add a smoothing topic, that allows to set an exponential smoothing factor,
  where 0 means perfect control and 1 means no control at all (don't move).

Commit:
Complete test suite for controller.

Commit:
Silence gcc warning on mismatching enumeral types.

Commit:
Make hold trajectory duration configurable.

Commit:
Implement hold traj without direction reversals.

- Implement hold trajectory (executed whenever a trajectory is cancelled, or the
  controller is started) without resulting in direction reversals.

- Implementation assumes that a segment going from (pos, vel) to (pos,-vel) is
  symmetric and has a zero-velocity point in the middle.

- Reduce default hold trajectory duration to 0.5s.

Commit:
Merge branch 'hydro-devel' of https://github.com/willowgarage/ros_controllers into joint_trajectory_controller_hydro

Commit:
Hydro compatibility patches.

jbohren commented:
Cool stuff, I'll look through this closely tomorrow. How hard do you think it would be to split the majority of the controller into a framework-independent library?

adolfo-rt commented:
The trajectory representation and sampling code is already in separate ros-agnostic headers. This accounts for all the core computations.

The code inside the actual controller is mostly for managing the ROS API, and just a little bit of it is ros_control-specific. It still amounts to some ~500 sloc.

Some extra work would be required to say, implement the ROS API in Orocos RTT, and using Orocos ports, lock-free buffers, the native actionlib server you've been working on, etc. It should be doable though, in the end we're talking mostly about methods for init, start, update and stop, plus callbacks.

davetcoleman commented:
I'll try to compile this now on Hydro...

davetcoleman commented:
For others - note that https://github.com/ros-controls/realtime_tools/pull/4 is required to build this

davetcoleman commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:
```diff
@@ -0,0 +1,664 @@
+  {
+    // Reject goal. Determine if the reason was invalid joints or something else
+    using internal::permutation;
+    std::vector<unsigned int> permutation_vector = permutation(joint_names_, gh.getGoal()->trajectory.joint_names);
```
Looks like you have a miss match of using a long unsigned int and just an unsigned int, but adding long caused other problems:

```
/home/dave/ros/ws_gazebo/src/ros_controllers/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:318:3:   instantiated from ‘bool joint_trajectory_controller::JointTrajectoryController<SegmentImpl, HardwareInterface>::init(hardware_interface::PositionJointInterface*, ro\
s::NodeHandle&, ros::NodeHandle&) [with SegmentImpl = trajectory_interface::QuinticSplineSegment<double>, HardwareInterface = hardware_interface::PositionJointInterface]’
/home/dave/ros/ws_gazebo/src/ros_controllers/joint_trajectory_controller/src/joint_trajectory_controller.cpp:59:1:   instantiated from here
/home/dave/ros/ws_gazebo/src/ros_controllers/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:527:114: error: conversion from ‘std::vector<long unsigned int>’ to non-scalar type ‘std::vector<unsigned int>’ requested
```

davetcoleman replied:
I tried to fix this myself for a while but couldn't figure out why it thinks the output is of type "long unsigned int"

davetcoleman committed:
Small CMake tweaks for catkin

Commit:
Use correct return type when calling permutation()

- Fixes a 32-64bit issue.

adolfo-rt commented:
@davetcoleman please try again, there's a new commit. Sorry for the inconvenience. This was most probably a 32-64 bit issue, where integral size types differ between architectures. The ultimate cause was me being lazy and not enforcing the correct return types.

As a sidenote, I can't wait until good C++11 support reaches stable Ubuntu distros (and ROS), to start coding [AAA-style](http://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto) and make issues like this one something of the past!

Commit:
Merge pull request #1 from davetcoleman/joint_trajectory_controller_hydro

Small CMake tweaks for catkin

davetcoleman commented:
I reviewed your fixes and tried to replicate them for this new set of similar errors I have, but I'm finding your code a little over-complex with all its typedefs, impl files and structure. Also, I'm just not putting much time into grasping it all. But here are my new errors after pulling in your fixes a few hours ago:

https://gist.github.com/davetcoleman/6309167

Commit:
Change return type of permutation() function.

Problem: permutation return type was vector<T>::size_type, which is:
- 32bits wide in 32bits Unix, which is the size of both unsigned int
  and unsigned long int.

- 64bits wide in 64bits Unix, which is the size of unsigned long int,
  but not unsigned int (the latter being 32bits wide).

I was doing at some points vector<unsigned int> = permutation(...),
which triggered a compile error in 64bit systems. Due to this, I'm
changing the return type of the function to be vector<unsigned int>,
and live with the implicit narrowing conversions that will take
place inside the implementation, since we won't be storing vectors
large enough to overflow an unsigned int.

adolfo-rt commented:
I installed gcc-multilib to run both 32 and 64 builds. I could reproduce the error, and should hopefully be fixed now. Hopefully this will land in the official Hydro release.

It's true that the code has lots of typedefs and typenames, as happens with all templated code. Speaking of the future, if we can manage to implement the controller without it being templated, some of the cruft will go away. This means 1. puligin-enabled segments, and 2. an easy way to connect to different hardware interfaces (could be a plugin as well).

@davetcoleman Thanks for the time you're putting into this.

davetcoleman commented:
It builds, well done! I also ran your tests (wow - you have a lot of good tests) and I got mixed results. Some of the tests failed and some passed, but even the passed ones have errors?

Results from Hydro Precise test run:
https://gist.github.com/davetcoleman/6324130

I'd say we can go ahead and merge this though.

@adolfo-rt today is the last day of my internship at OSRF so I won't be around as much - need to work on papers :)

davetcoleman commented on joint_trajectory_controller/CMakeLists.txt:
```diff
@@ -0,0 +1,140 @@
+      trajectory_msgs
+      rostest
+      controller_manager
+      xacro
```
I've been having issues depending on xacro in groovy... doesn't seem to work. Probably not a direct issue here but I have to remove these references to compile. What do you think @wjwwood ?

wjwwood replied:
I think if you are having trouble with `xacro` on Groovy you should file an issue against `xacro`, it should work as far as I know.

wjwwood replied:
Well, also this was 6 months ago :anguished:, so ignore the noise.

adolfo-rt commented:
All failing tests are rostests. It seems to be a setup issue. Apparently not all nodes from the .test file are being launched, eg. [this line](https://gist.github.com/davetcoleman/6324130#file-joint_trajectory_controller_test-L158) is failing on a precondition that indicates that the controller to test is not running (or maybe taking too long to start). I'll see if I can sort this out on my end.

Don't worry about passing tests containing ERROR log statements. It's expected behavior, as the test is exercising that it can gracefully handle invalid/inconsistent input.

Commit:
Comment xacro dependency as it breaks the build.

TODO: Figure out what's going on here?.

adolfo-rt commented:
I setup a Groovy installation and verified that the xacro dependency causes problems, so I commented it for now. Other than that, I still have to figure out how to setup rostests, which are not working. Let's see if I can fix it on time.

Commit:
Fix remaining Hydro build and test issues.

adolfo-rt commented:
I just tested the latest changes in a hydro build, and all tests are passing. I'm giving this a go.

davetcoleman commented:
Well done!

jbohren commented:
@adolfo-rt Can you give some more details on the trajectory reversal behavior that you mentioned in 129ea1b?

adolfo-rt commented:
If a joint currently has a state (pos_1, vel_1), where the velocity vel_1 is nonzero, and you request the joint to stop by sending the command (pos_1, 0) in time t, what will happen is that the joint will move past pos_1 reverse its direction of motion, and then go back and reach pos_1, this time with zero velocity. I believe this is what you had in your original Reflexxes based implementation.

What 129ea1b does is to compute a stop trajectory starting at state (pos_1, vel_1) and with duration t that will stop past pos_1, but without direction reversals. In particular, it generates a segment going from (pos_1, vel_1) to (pos_1, -vel_1) with duration 2t, and then samples it at the midpoint to get pos_2, where (assuming symmetry) the velocity is zero. The stop trajectory then goes from (pos_1, vel_1) to (pos_2, 0) in time t.

The limitation of this approach is that it assumes symmetry in the segment from (pos_1, vel_1) to (pos_1, -vel_1). It is the case for the current spline representation, but might not be the case for other representations, e.g., a Reflexxes based implementation might yield non-symmetric segments when joint limit violations are involved.

Notice that by default the stop trajectory duration is zero, and the deceleration profile is determined by either low-level joint limits enforcing (outside the controller) or by the hardware itself.

jbohren commented:
> that will stop _past_ pos_1, but without direction reversals

Ah alright. That's what I was wondering about.

> Notice that by default the stop trajectory duration is zero, and the deceleration profile is determined by either low-level joint limits enforcing (outside the controller) or by the hardware itself.

I did notice this. Have you put any time into creating a segment type that incorporates joint velocity limits?

adolfo-rt commented:
> > Notice that by default the stop trajectory duration is zero, and the deceleration profile is determined by either low-level joint limits enforcing (outside the controller) or by the hardware itself.
> 
> I did notice this. Have you put any time into creating a segment type that incorporates joint velocity limits? 

No, I ran out of time budget when coding the controller. I was pondering whether it made more sense to put it here in the controller or defer limits enforcing to the robot hardware abstraction. The upside of having it in the controller is that effort-controlled robots would benefit from it as well.

Wherever it ended up being implemented, I really liked the idea of combining Reflexxes with the  implementation in JointLimitsInterface. Since Reflexxes does not enforce position limits, we could adjust online the velocity limits depending on how far we are from the position limits.
